### PR TITLE
#342 clarify the circuit breaker rolling window

### PR DESCRIPTION
**Pull request by Emily-Jiang**
Signed-off-by: Emily Jiang <emijiang@uk.ibm.com>

### CODE REVIEW

File: tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/circuitbreaker/clientserver/CircuitBreakerClientRollingWindow.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+ * A client to exercise Circuit Breaker thresholds, with a SuccessThreshold of 2,
+ * a requestVolumeThreshold of 4, failureRatio of 0.5 and a 1 millisecond delay
+ *
+ * @author <a href="mailto:emijiang@uk.ibm.com">Emily Jiang</a>
+ */
+@RequestScoped
+public class CircuitBreakerClientRollingWindow implements Serializable {
+        
+    private static final long serialVersionUID = 1L;
+    private int counterForInvokingServiceA = 0;
```
Comment (Azquelt) [inline tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/circuitbreaker/clientserver/CircuitBreakerClientRollingWindow.java]: Issue #323 requested that we don't reuse circuit breakers between tests. (We were expecting each test to have it's own new request scope and for that to mean we got a new circuit breaker instance).

I wonder if that means we should avoid relying on each test having a fresh request scope? If so we should have one counter for each test method here. (or we should reset the counter to zero at the start of each test, but that seems more prone to error)

File: tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/CircuitBreakerTest.java [hunk 2]
```diff
@@ -624,4 +627,98 @@ public void testClassLevelCircuitOverrideNoDelay() {
...
+            catch (Exception ex) {
+                // Not Expected
+                Assert.fail("serviceA should throw a RuntimeException or CircuitBreakerOpenException in testRollingWindowCircuitOpen2 "
+                        + "on iteration " + i);
+            }
+        }
+        int serviceAExecutions = clientForRW.getCounterForInvokingServiceA();
+        Assert.assertTrue(cbExceptionThrown, "The CircuitBreaker exception in testRollingWindowCircuitOpen2 was not thrown correctly.");
+
+        Assert.assertEquals(serviceAExecutions, 6, "The number of executions should be 6");
```
Comment (Azquelt) [inline tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/CircuitBreakerTest.java]: I think we expect 5 executions here, the sixth throws a CircuitBreakerOpenException and the method body is not executed.

File: tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/CircuitBreakerTest.java [hunk 3]
```diff
@@ -624,4 +627,98 @@ public void testClassLevelCircuitOverrideNoDelay() {
...
+            catch (Exception ex) {
+                // Not Expected
+                Assert.fail("serviceA should throw a RuntimeException or CircuitBreakerOpenException in testRollingWindowCircuitOpen "
+                     + "on iteration " + i);
+            }
+        }
+        int serviceAExecutions = clientForRW.getCounterForInvokingServiceA();
+        Assert.assertTrue(cbExceptionThrown, "The CircuitBreaker exception in testRollingWindowCircuitOpen was not thrown correctly.");
+
+        Assert.assertEquals(serviceAExecutions, 5, "The number of executions should be 5");
```
Comment (Azquelt) [inline tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/CircuitBreakerTest.java]: I think we expect 4 executions here, the fifth throws a CircuitBreakerOpenException and the method body is not executed.

File: spec/src/main/asciidoc/circuitbreaker.asciidoc [hunk 4]
```diff
@@ -38,7 +38,37 @@ succeed before the circuit can be closed. After the specified number of successf
...
+   return conn;
+}
+----
+Below are a couple scenarios to demonstrate when the circuit opens for the above code snippet.
+* Scenario 1
+     Request 1 - success
+     Request 2 - failure
+     Request 3 - failure
+     Request 4 - success
+     Request 5 - CircuitOpenException
```
Comment (Azquelt) [inline spec/src/main/asciidoc/circuitbreaker.asciidoc]: Exception name is `CircuitBreakerOpenException`. Also below.

Comment (Azquelt) [inline spec/src/main/asciidoc/circuitbreaker.asciidoc]: This section does not render correctly in the spec pdf

File: tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/circuitbreaker/clientserver/CircuitBreakerClientRollingWindow.java (hunk 1 above)
Comment (Emily-Jiang) [reply tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/circuitbreaker/clientserver/CircuitBreakerClientRollingWindow.java]: ah. My c+p mistake. I should have used two different counters. fixed now.

File: tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/CircuitBreakerTest.java (hunk 2 above)
Comment (Emily-Jiang) [reply tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/CircuitBreakerTest.java]: yep. thanks.

File: tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/CircuitBreakerTest.java (hunk 3 above)
Comment (Emily-Jiang) [reply tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/CircuitBreakerTest.java]: yep. thanks

File: spec/src/main/asciidoc/circuitbreaker.asciidoc (hunk 4 above)
Comment (Emily-Jiang) [reply spec/src/main/asciidoc/circuitbreaker.asciidoc]: It seems my asciidoc preview rendered ok but not the generated one. fixed.

File: tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/circuitbreaker/clientserver/CircuitBreakerClientRollingWindow.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+    public int getCounterForInvokingService1() {
+           return counterForInvokingService1;
+    }
+
+    public int getCounterForInvokingService2() {
+        return counterForInvokingService2;
+ }
+
+    @CircuitBreaker(successThreshold = 2, requestVolumeThreshold = 4, failureRatio=0.5, delay = 1)
```
[the remark this answers is not in the document]
Comment (Emily-Jiang) [reply tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/circuitbreaker/clientserver/CircuitBreakerClientRollingWindow.java]: removed

File: tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/circuitbreaker/clientserver/CircuitBreakerClientRollingWindow.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        if((counterForInvokingService1 ==2 ) || (counterForInvokingService1 ==3)) {
+                throw new RuntimeException("Connection failed");
+        }
+        return null;
+    
+    
+    }
+    //simulate a backend service
+    //Throw exception for the 2nd and 5th request.
+    @CircuitBreaker(successThreshold = 2, requestVolumeThreshold = 4, failureRatio=0.5, delay = 1)
```
[the remark this answers is not in the document]
Comment (Emily-Jiang) [reply tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/circuitbreaker/clientserver/CircuitBreakerClientRollingWindow.java]: fixed. thanks

File: tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/circuitbreaker/clientserver/CircuitBreakerClientRollingWindow.java (hunk 1 above)
Comment (brunobat) [reply tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/circuitbreaker/clientserver/CircuitBreakerClientRollingWindow.java]: That private counterForInvokingServiceA attribute is counting calls to the same bean instance. 
Different requests scoped should see different CircuitBreakerClientDefaultSuccessThreshold objects.
In order to be effective that attribute would need to be static.
If we are seeing a non static attribute being incremented, it means its being called from the same context.
This makes me think the all the requests are being made from the same scope making the test ineffective. 
What do you guys think?
In order to be

Comment (Azquelt) [reply tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/circuitbreaker/clientserver/CircuitBreakerClientRollingWindow.java]: @brunobat I don't think I'm quite following you. You're correct in that the counter will only count calls to the same instance, but the test is making all of its calls against a single instance, so I'd expect it to work. Within one test method, I'd only expect there to be one request context, so we'd only get one instance of the request scoped bean.

The original code assumed that each test method will have a different request context as it relied on each test using a different instance to ensure that the counter was zero at the start of the test. This is true in most cases, but some embedded containers will run all tests in the same request context.

In the current code, `counterForInvokingService1` counts invocations of `service1RollingWindowOpenAfter4()`. `testRollingWindowCircuitOpen()` makes five requests to this method against the same instance of `CircuitBreakerClientRollingWindow`, and then checks the counter.

Comment (brunobat) [reply tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/circuitbreaker/clientserver/CircuitBreakerClientRollingWindow.java]: Ok, I was mixing `@RequestScoped` with `@ApplicationScoped` In this case the test uses `@RequestScope` as we have been doing.

**Comment by Emily-Jiang** [comment]
this  closes #342

**Pull request merged**